### PR TITLE
object: hotfix for object head method (header stringer)

### LINKED ISSUE #34: object.Head RPC with full-headers flag leads to panic
Affected version: `neofs-cli v0.7.0`

It seems like stringer for object header is completely broken now:

```
$ ./bin/neofs-cli --config .neofs-cli.yml.remote object head --full-headers --cid 4ZJydjUYwnrHEDPBhfNE1gmNNRPfggTjPSUXj6xgEXY1 --oid 163beb35-57f5-4758-b771-d5efd8d68a4f
System headers:
  Object ID   : 163beb35-57f5-4758-b771-d5efd8d68a4f
  Owner ID    : AK2nJJpJr6o664CWJKi1QRXjqeic2zRp8y
  Container ID: 4ZJydjUYwnrHEDPBhfNE1gmNNRPfggTjPSUXj6xgEXY1
  Payload Size: 131.701KB
  Version     : 1
  Created at  : epoch #20, 2020-04-29 17:50:10 +0300 MSK
Other headers:
panic: invalid Go type hash.Hash for field object.Header.HomoHash

goroutine 1 [running]:
google.golang.org/protobuf/internal/impl.newSingularConverter(0xe265c0, 0xc8a420, 0xe23220, 0xc00020c540, 0x1, 0x0)
        /home/alexvanin/Project/NeoFS/neofs-cli/vendor/google.golang.org/protobuf/internal/impl/convert.go:143 +0xff5
google.golang.org/protobuf/internal/impl.NewConverter(0xe265c0, 0xc8a420, 0xe23220, 0xc00020c540, 0x0, 0x0)
        /home/alexvanin/Project/NeoFS/neofs-cli/vendor/google.golang.org/protobuf/internal/impl/convert.go:60 +0xd3
google.golang.org/protobuf/internal/impl.fieldInfoForOneof(0xe23220, 0xc00020c540, 0xb84da3, 0x5, 0x0, 0x0, 0xe265c0, 0xc23c20, 0xb84daa, 0x16, ...)
        /home/alexvanin/Project/NeoFS/neofs-cli/vendor/google.golang.org/protobuf/internal/impl/message_reflect_field.go:42 +0x1bd
google.golang.org/protobuf/internal/impl.(*MessageInfo).makeKnownFieldsFunc(0xc0004228c0, 0x28, 0xffffffffffffffff, 0x10, 0xffffffffffffffff, 0xc000518a50, 0xc000518a80, 0xc000518ab0, 0xc000518ae0)
        /home/alexvanin/Project/NeoFS/neofs-cli/vendor/google.golang.org/protobuf/internal/impl/message_reflect.go:57 +0x33a
google.golang.org/protobuf/internal/impl.(*MessageInfo).makeReflectFuncs(0xc0004228c0, 0xe265c0, 0xc7c940, 0x28, 0xffffffffffffffff, 0x10, 0xffffffffffffffff, 0xc000518a50, 0xc000518a80, 0xc000518ab0, ...)
        /home/alexvanin/Project/NeoFS/neofs-cli/vendor/google.golang.org/protobuf/internal/impl/message_reflect.go:36 +0x63
google.golang.org/protobuf/internal/impl.(*MessageInfo).initOnce(0xc0004228c0)
        /home/alexvanin/Project/NeoFS/neofs-cli/vendor/google.golang.org/protobuf/internal/impl/message.go:91 +0x182
google.golang.org/protobuf/internal/impl.(*MessageInfo).init(0xc0004228c0)
        /home/alexvanin/Project/NeoFS/neofs-cli/vendor/google.golang.org/protobuf/internal/impl/message.go:73 +0x3c
google.golang.org/protobuf/internal/impl.(*messageReflectWrapper).WhichOneof(0xc0000fa0f0, 0xe1e1c0, 0xc00023a480, 0xc00020c000, 0xc40a00)
        /home/alexvanin/Project/NeoFS/neofs-cli/vendor/google.golang.org/protobuf/internal/impl/message_reflect_gen.go:233 +0x33
github.com/golang/protobuf/proto.(*textWriter).writeMessage(0xc000518810, 0xe20280, 0xc0000fa0f0, 0x0, 0x0)
        /home/alexvanin/Project/NeoFS/neofs-cli/vendor/github.com/golang/protobuf/proto/text_encode.go:273 +0x96b
github.com/golang/protobuf/proto.(*TextMarshaler).marshal(0x13582c6, 0xe10560, 0xc000001980, 0x1, 0x1, 0xf, 0x0, 0x0)
        /home/alexvanin/Project/NeoFS/neofs-cli/vendor/github.com/golang/protobuf/proto/text_encode.go:86 +0x18a
github.com/golang/protobuf/proto.(*TextMarshaler).Text(...)
        /home/alexvanin/Project/NeoFS/neofs-cli/vendor/github.com/golang/protobuf/proto/text_encode.go:44
github.com/golang/protobuf/proto.CompactTextString(...)
        /home/alexvanin/Project/NeoFS/neofs-cli/vendor/github.com/golang/protobuf/proto/text_encode.go:106
github.com/nspcc-dev/neofs-api-go/object.(*Header).String(...)
        /home/alexvanin/Project/NeoFS/neofs-cli/vendor/github.com/nspcc-dev/neofs-api-go/object/types.pb.go:231
main.head(0xc00031a500, 0x5, 0x8)
        /home/alexvanin/Project/NeoFS/neofs-cli/object.go:261 +0x9e1
github.com/urfave/cli/v2.(*Command).Run(0xc00013f200, 0xc000167080, 0x0, 0x0)
        /home/alexvanin/Project/NeoFS/neofs-cli/vendor/github.com/urfave/cli/v2/command.go:161 +0x4e0
github.com/urfave/cli/v2.(*App).RunAsSubcommand(0xc000682480, 0xc000166d80, 0x0, 0x0)
        /home/alexvanin/Project/NeoFS/neofs-cli/vendor/github.com/urfave/cli/v2/app.go:433 +0xa9a
github.com/urfave/cli/v2.(*Command).startApp(0xc00013e5a0, 0xc000166d80, 0x7fffde63c067, 0x6)
        /home/alexvanin/Project/NeoFS/neofs-cli/vendor/github.com/urfave/cli/v2/command.go:275 +0x67f
github.com/urfave/cli/v2.(*Command).Run(0xc00013e5a0, 0xc000166d80, 0x0, 0x0)
        /home/alexvanin/Project/NeoFS/neofs-cli/vendor/github.com/urfave/cli/v2/command.go:91 +0xa12
github.com/urfave/cli/v2.(*App).RunContext(0xc000682300, 0xe14b60, 0xc0001302e8, 0xc000138000, 0xa, 0xa, 0x0, 0x0)
        /home/alexvanin/Project/NeoFS/neofs-cli/vendor/github.com/urfave/cli/v2/app.go:302 +0x814
github.com/urfave/cli/v2.(*App).Run(...)
        /home/alexvanin/Project/NeoFS/neofs-cli/vendor/github.com/urfave/cli/v2/app.go:211
main.main()
        /home/alexvanin/Project/NeoFS/neofs-cli/main.go:22 +0x25c
```

New object stringer will be available with later `neofs-api-go` release (see https://github.com/nspcc-dev/neofs-api-go/pull/68), but current `neofs-api-go` and `neofs-node` versions are LTS, so we need to fix this.

### LINKED ISSUE #36: "sg get" and "sg delete" commands require "oid" parameter
Storage groups stored as an objects in NeoFS. To get info about storage group CLI makes `head` request on storage group object. 

Command `neofs-cli sg get` requires `--sgid` parameter, but it reuses flags from `neofs-cli object head` command, where `--oid` parameter is mandatory. 

Therefore we got an error:
```
$ ./bin/neofs-cli sg get --cid 4ZJydjUYwnrHEDPBhfNE1gmNNRPfggTjPSUXj6xgEXY1 --sgid 612a88eb-a183-4676-9877-9aadfa581faf
NAME:
   neofs-cli sg get - get storage group from the system

USAGE:
   get --cid <cid> --sgid <sgid> 

DESCRIPTION:
   get information about existing storage group

OPTIONS:
   --cid value     user container ID
   --oid value     object ID to receive
   --full-headers  return all headers (default: false)
   --raw           send request with raw flag set (default: false)
   --sgid value    storage group id
   --help, -h      show help (default: false)
   
Required flag "oid" not set
```

### PR DESCRIPTION
**Pull request by im-kulikov**
closes #34 
closes #36

Example

```
→ go run ./ object head --full-headers --cid 4ZJydjUYwnrHEDPBhfNE1gmNNRPfggTjPSUXj6xgEXY1 --oid ab736579-0e80-4fc8-8448-29425c6d3be3
Object:
	SystemHeader:
		- ID=7e0b9c6c-aabc-4985-949e-2680e577b48b
		- CID=11111111111111111111111111111111
		- OwnerID=ALYeYC41emF6MrmUMc4a8obEPdgFhq9ran
		- Version=1
		- PayloadLength=1
		- CreatedAt={UnixTime=1 Epoch=1}
	ExtendedHeaders:
		- Type=Link
		  Value={Type=Child ID=7e0b9c6c-aabc-4985-949e-2680e577b48b}
		- Type=Redirect
		  Value={CID=11111111111111111111111111111111 OID=7e0b9c6c-aabc-4985-949e-2680e577b48b}
		- Type=UserHeader
		  Value={Key=test_key Val=test_value}
		- Type=Transform
		  Value=Split
		- Type=Tombstone
		  Value=MARKED
		- Type=Verify
		  Value={PublicKey=010203040506 Signature=010203040506}
		- Type=HomoHash
		  Value=00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
		- Type=PayloadChecksum
		  Value=010203040506
		- Type=Integrity
		  Value={Checksum=010203040506 Signature=010203040506}
		- Type=StorageGroup
		  Value={DataSize=5 Hash=00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000 Lifetime={Unit=UnixTime Value=555}}
		- Type=PublicKey
		  Value=010203040506
	Payload: []byte{0x1, 0x2, 0x3, 0x4, 0x5, 0x6, 0x7}
```

---

```
→ go run ./ sg get -h
NAME:
   neofs-cli sg get - get storage group from the system

USAGE:
   get --cid <cid> --sgid <sgid> 

DESCRIPTION:
   get information about existing storage group

OPTIONS:
   --cid value   user container ID
   --raw         send request with raw flag set (default: false)
   --sgid value  storage group id
   --help, -h    show help (default: false)

→ go run ./ sg delete -h
NAME:
   neofs-cli sg delete - delete storage group from the system

USAGE:
   delete --cid <cid> --sgid <sgid>

DESCRIPTION:
   delete user's storage group

OPTIONS:
   --cid value   user container ID
   --raw         send request with raw flag set (default: false)
   --sgid value  storage group id
   --help, -h    show help (default: false)
```

### CODE REVIEW

File: object.go [hunk 1]
```diff
@@ -258,7 +258,23 @@ func head(c *cli.Context) error {
...
+			// todo replace all this with object.Stringify in future
+			switch hdr := resp.Object.Headers[i].Value.(type) {
+			case *object.Header_HomoHash:
+				fmt.Printf("  homohash = %s\n", hdr.HomoHash.String())
+			case *object.Header_PublicKey:
+				fmt.Printf("  pubkey = %02x\n", hdr.PublicKey.Value)
+			case *object.Header_PayloadChecksum:
+				fmt.Printf("  payload_checksum = %02x\n", hdr.PayloadChecksum)
+			case *object.Header_Integrity:
+				fmt.Printf("  integrity = %s\n", hdr.Integrity.String())
```
[the remark this answers is not in the document]
Comment (im-kulikov) [reply object.go:270]: kk

**Comment by im-kulikov** [comment]
ported `object.Stringify` from `neofs-api-go`

File: object.go [hunk 2]
```diff
@@ -248,17 +248,127 @@ func head(c *cli.Context) error {
...
-	if len(resp.Object.Headers) != 0 {
-		fmt.Println("Other headers:")
-		for i := range resp.Object.Headers {
-			fmt.Println("  " + resp.Object.Headers[i].String())
+	return objectStringify(os.Stdout, resp.Object)
+}
+
+// objectStringify converts object into string format.
+func objectStringify(dst io.Writer, obj *object.Object) error {
+	// put empty line
```
Comment (AlexVanin) [inline object.go:256]: Why do we need empty line there? I think it would be cleaner if object line will be first output line.

File: object.go [hunk 3]
```diff
@@ -248,17 +248,127 @@ func head(c *cli.Context) error {
...
+
+		}
+
+		if _, err := fmt.Fprintf(dst, "\t\t- %s=%v\n", key, val); err != nil {
+			return err
+		}
+	}
+
+	// put user headers
+	if _, err := fmt.Fprintln(dst, "\tUserHeaders:"); err != nil {
```
Comment (AlexVanin) [inline object.go:296]: It is not user headers, it is `ExtendedHeaders`

Comment (im-kulikov) [reply object.go]: done

File: storagegroup.go [hunk 4]
```diff
@@ -49,7 +71,7 @@ func listSG(c *cli.Context) error {
 
 func getSG(c *cli.Context) error {
 	if err := c.Set(objFlag, c.String(sgidFlag)); err != nil {
-		return err
+		return errors.Wrap(err, "HERE!!!")
```
Comment (AlexVanin) [inline storagegroup.go:74]: Remove debug message.

Comment (im-kulikov) [reply storagegroup.go]: done

**Pull request merged**